### ISSUE TITLE
mask_privacy efficiency

### ISSUE THREAD

**Issue by Mr-Dave** [Motion-Project/motion#340]
The following is moved over from the mailing list as sent by @pascal-fb-martin

While searching for something else, I found that function mlp_mask_privacy (in file motion.c) executes a slow double loop with a lot of invariant re-calculation in the inner loop:

      for (indxrow = 0; indxrow < cnt->imgs.height; indxrow++) {
          for (indxcol = 0; indxcol < cnt->imgs.width; indxcol++) {
              cnt->current_image->image[indxcol + (indxrow*cnt->imgs.width)] |= cnt->imgs.mask_privacy[indxcol + (indxrow*cnt->imgs.width)];
          }
      }

The operations are for "unsigned char". The current_image and mask_privacy have the same layout: a flat array of bytes. A single loop would do at a lower cost. What tickled me was the code comment: "This is done strictly for processing efficiency to lower cpu". Oh, well..

The following code is equivalent, but more efficient:

      int index = cnt->imgs.height*cnt->imgs.width;
      unsigned char *image = cnt->current_image->image;
      const unsigned char *mask = cnt->imgs.mask_privacy;

      while (index > sizeof(unsigned long)) {
          *((unsigned long *)image) |= *((unsigned long *)mask);
          image += sizeof(unsigned long);
          mask += sizeof(unsigned long);
          index -= sizeof(unsigned long);
      }
      while (--index >= 0) { // process the last few remaining bytes, if any.
          *(image++) |= *(mask++);
      }

The relative results (with or without gcc optimizations) are below:

    $ cc imagespeed.c   -o imagespeed
    $ ./imagespeed 10000
    Testing privacy mask loop (repeat 10000 times)
    Execution time (original loop): 25 sec.
    Execution time (improved loop): 3 sec.
    $ cc -O2 imagespeed.c   -o imagespeed
    $ ./imagespeed 100000
    Testing privacy mask loop (repeat 100000 times)
    Execution time (original loop): 64 sec.
    Execution time (improved loop): 4 sec.

Doing a flat loop with minimal calculations about double the speed. Preceding it with the "unsigned long" loop multiply the speed by an additional factor of 8 on my 64 bit PC. To my surprise, the gain is much lower (but still dramatic) when compiled without optimization.

I attached the source file of my performance test. It simulates the same struct indirections as used by motion.c and starts by generating a binary pattern for the current image and another for the mask, calculates the masked image using each method and compare the binary result. If the check passed, it runs the performance tests. That is the full extend of my testing thus far.

Comments are welcome. I have prepared a modified version of motion.c, but I still have to test it.

Thanks.

Pascal.

**Comment by Mr-Dave** [comment]
Note that the code for this particular function has become "even worse".   

The version above was the first variant that was only dealing with the y color space.  The "more efficient" comment was really meant to be addressing the use of the white color as the omitted section instead of black.  It permitted the use of the <code> |= </code> instead of needing a <code>if</code>.  

Unfortunately, the need for a <code>if</code> arose with the need to handle the cr and cb color spaces.  The byte transformation for the cr and cb needs to be 0x80 instead of 0xFF.  

Any more efficient method for this is welcome.

**Comment by pascal-fb-martin** [comment]
Regarding the |= operation: what would be wrong with not inverting the
privacy mask and using &= instead?

Sorry for not detecting the recent changes. I will checkout the new code
and analyze the thing again.

Thanks.

**Comment by pascal-fb-martin** [comment]
The latest motion code is actually more efficient (as it removes the double loop), even if more expensive overall, but is still not too complicated to improve. The code is becoming more complicated, through (if you look closely, it is a similar approach to the MMX motion detection code: process 8 bytes at a time :smile:):

(Code below was not tested, only the performance test variant--see updated file attached 
[imagespeed.zip](https://github.com/Motion-Project/motion/files/820520/imagespeed.zip).)
```
   unsigned char *image = cnt->current_image->image;
   const unsigned char *mask = cnt->imgs.mask_privacy;
   int pixels = cnt->imgs.height * cnt->imgs.width;

   // Mask brightness.
   //
   int index = pixels;

   while (index >= sizeof(unsigned long)) {
      *((unsigned long *)image) &= *((unsigned long *)mask);
      image += sizeof(unsigned long);
      mask += sizeof(unsigned long);
      index -= sizeof(unsigned long);
   }
   while (--index >= 0) {
      *(image++) &= *(mask++);
   }

   // Mask chrominance.
   //
   index = cnt->imgs.size - pixels;
   const unsigned char *mask2 = cnt->imgs.mask_privacy_or + pixels;

   while (index >= sizeof(unsigned long)) {
      *((unsigned long *)image) &= *((unsigned long *)mask);
      *((unsigned long *)image) |= *((unsigned long *)mask2);
      image += sizeof(unsigned long);
      mask += sizeof(unsigned long);
      mask2 += sizeof(unsigned long);
      index -= sizeof(unsigned long);
   }
   while (--index >= 0) {
      if (! *(mask++)) *(image++) = 0x80; // Mask last remaining bytes.
   }
```

The cost is to build mask_privacy_or (more MB wasted..):
```
if (cnt->imgs.mask_privacy[x] == 0x00)
   mask_privacy_or[x] = 0x80;
else
   mask_privacy_or[x] = 0x00;
```
In other words, where mask_privacy is used to clear masked bits, mask_privacy_or is used to set bit 7.

The new results are still good (improved speed is still significantly faster than the old original):

```
$ cc imagespeed.c   -o imagespeed
$ ./imagespeed 10000
Testing privacy mask loop (repeat 10000 times)
Execution time (original loop): 28 sec.
Execution time (improved loop): 5 sec.
$ cc -O2 imagespeed.c   -o imagespeed
$ ./imagespeed 100000
Testing privacy mask loop (repeat 100000 times)
Execution time (original loop): 97 sec.
Execution time (improved loop): 16 sec.
```

**Comment by pascal-fb-martin** [comment]
The following code might be a bit easier on the compiler's optimizer (but the results show some performance instability--is my PC telling me to go to bed?):

```
      *((unsigned long *)image) &= *((unsigned long *)mask);
      mask += sizeof(unsigned long);
      *((unsigned long *)image) |= *((unsigned long *)mask2);
      image += sizeof(unsigned long);
      mask2 += sizeof(unsigned long);
```

**Comment by pascal-fb-martin** [comment]
Hoops: stupid bug in previous version of optimized code.

The code:
```
if (! *(mask++)) *(image++) = 0x80; // Mask last remaining bytes.
```
Should have been:
```
if (! *(mask++)) *image = 0x80; // Mask last remaining bytes.
image += 1;
```
Comment above corrected.

**Comment by Mr-Dave** [comment]
This is a significant improvement!  I tested the improved function in the Motion code and it masked everything great.
